Cython/Utility/Optimize.c

```diff
@@ -361,8 +361,20 @@ static CYTHON_INLINE int __Pyx_dict_iter_next(
             }
             Py_INCREF(key);
             Py_INCREF(value);
+            #if CYTHON_ASSUME_SAFE_MACROS
             PyTuple_SET_ITEM(tuple, 0, key);
             PyTuple_SET_ITEM(tuple, 1, value);
+            #else
+            if (unlikely(PyTuple_SetItem(tuple, 0, key) < 0)) {
+                Py_DECREF(value); // we haven't set this yet
```

Comment:
We haven't actually set `key` in this case either. We need to decref both (and `value` in the failure case below).

Reply:
`PyTuple_SetItem` decrefs `key` on failure (https://github.com/python/cpython/blob/939fc6d6eab9b7ea8c244d513610dbdd556503a7/Objects/tupleobject.c#L122C10-L122C10) so that it effectively steals a reference whether or not it succeeds.

I'll add a comment because it isn't obvious though